### PR TITLE
Improve music module

### PR DESCRIPTION
**Pull request by GTG3000**
Added support for playlists and downloading songs before playing.

### CODE REVIEW

File: plugins/command_dispatcher.py [hunk 1]
```diff
@@ -103,8 +103,9 @@ def deregister(self, fn, name, is_alias=False):
                               f"**WARNING: {err} ANALYSIS: Proper usage: {deco}{command} {fn.syntax}.**")
             else:
                 await respond(self.client, data, f"**WARNING: {err}.**")
-        except PermissionError:
-            await respond(self.client, data, "**NEGATIVE. INSUFFICIENT PERMISSION: <usernick>.**")
+        except PermissionError as e:
+            err = f"\nANALYSIS: {e}" if e else ""
+            await respond(self.client, data, f"**NEGATIVE. INSUFFICIENT PERMISSION: <usernick>.{err}**")
```
Comment (medeor413) [inline plugins/command_dispatcher.py]: Add a space here between the period and `{err}`.

File: plugin_manager.py [hunk 2]
```diff
@@ -156,7 +156,11 @@ def final_load(self):
             time = 60
         await asyncio.sleep(time)
         self.logger.debug("Writing to shelve...")
-        self.shelve.sync()
+        try:
+            self.shelve.sync()
+        except Exception as e:
+            err = f"{e}." if e else ""
+            self.logger.error(f"Error writing to shelve. {err}")
```
Comment (medeor413) [inline plugin_manager.py]: Better to `self.logger.exception("Error writing to shelve. ", exc_info=True)` here.

File: plugins/music.py [hunk 3]
```diff
@@ -24,15 +29,21 @@ class MusicPlayer(BasePlugin):
         'max_queue_length': 30,
         'default_volume': 15,
         'allow_pause': True,
+        'allow_playlists': True,
+        'twich_stream': False,
```
Comment (medeor413) [inline plugins/music.py]: twitch*

File: plugins/music.py [hunk 4]
```diff
@@ -24,15 +29,21 @@ class MusicPlayer(BasePlugin):
...
+        'download_songs': True,
+        'download_songs_timeout': 259200,
```
Comment (medeor413) [inline plugins/music.py]: Why this number?

File: plugins/music.py [hunk 5]
```diff
@@ -74,23 +85,27 @@ class MusicPlayer(BasePlugin):
         self.volume = c.default_volume
         self.max_length = c.max_video_length
         self.max_queue = c.max_queue_length
+        self.ytdl_options["playlistend"] = self.max_queue+5
         self.m_channel = c.music_channel if c.force_music_channel else False
         self.allow_pause = c.allow_pause
+        self.stream = c.twich_stream
```
Comment (medeor413) [inline plugins/music.py]: twitch*

File: plugins/music.py [hunk 6]
```diff
@@ -429,14 +460,215 @@ class MusicPlayer(BasePlugin):
...
             raise SyntaxError("Expected arguments!")
 
+    @Command("delqueue",
+             category="music",
+             doc="Erases serialised queue. Useful if bot was shut down with ridiculous queue saved.",
+             perms="mute_members")
+    async def _delqueue(self, data):
+        if self.check_ban(data.author.id):
+            raise PermissionError("You are banned from using the music module.")
+        await respond(self.client, data, "**AFFIRMATIVE. Purting.**")
```
Comment (medeor413) [inline plugins/music.py]: "Purting?"

File: plugins/music.py [hunk 7]
```diff
@@ -429,14 +460,215 @@ class MusicPlayer(BasePlugin):
...
+        self.storage["serialized_queue"] = []
+
     # Music playing
 
+    """
```
Comment (medeor413) [inline plugins/music.py]: Comment blocks aren't indented

File: plugins/music.py [hunk 8]
```diff
@@ -509,33 +779,59 @@ class MusicPlayer(BasePlugin):
...
+            self.queue.append(t_player)
 
     def start_timer(self, loop):
         asyncio.set_event_loop(loop)
         try:
             loop.run_until_complete(self.display_time())
-        except:
-            pass
+        except Exception as e:
+            self.logger.error(f"ERROR {e}")
```
Comment (medeor413) [inline plugins/music.py]: Use `self.logger.exception("Error starting timer. ", exc_info=True)`; always include what function is erroring in an error message.

File: plugins/music.py [hunk 9]
```diff
@@ -509,33 +779,59 @@ class MusicPlayer(BasePlugin):
...
+                    if time.time() - time_added > self.plugin_config["download_songs_timeout"]:
+                        del_list.append(song)
+                for song in del_list:
+                    try:
+                        os.remove(song)
+                        self.storage["stored_songs"].pop(song)
+                        print(f"Attempting to remove {song}.")
+                    except Exception as e:
+                        err = e if e else "Some error happened."
+                        self.logger.error(err)
```
Comment (medeor413) [inline plugins/music.py]: Same as above

File: plugins/command_dispatcher.py (hunk 1 above)
Comment (GTG3000) [inline plugins/command_dispatcher.py]: There isn't a space because err starts with a newline if it's a thing.

File: plugins/music.py (hunk 4 above)
Comment (GTG3000) [inline plugins/music.py]: Because three days.

**Pull request merged**